### PR TITLE
Vue3 Upgrade

### PR DESCRIPTION
**Pull request by sha1sum**
This upgrades to Vue 3 with all the accompanying changes necessary.

### CODE REVIEW

**Comment by MaxLeiter** [comment]
@xPaw tagging you here in case you wanted to take a look because I know you were at one point working on this

**Comment by xPaw** [comment]
Well my old branch is still up at https://github.com/thelounge/thelounge/tree/vue3 if you want to take a look whether anything was different on it.

**Comment by sha1sum** [comment]
@MaxLeiter @xPaw we are ready to go! This obviously deserves a pull down and check for yourselves, but in my opinion, LGTM 👍

File: webpack.config-test.js [hunk 1]
```diff
@@ -36,14 +36,18 @@ config.optimization.splitChunks = false;
 config.plugins = [
 	new VueLoaderPlugin(),
 
+	new webpack.DefinePlugin({
+		__VUE_OPTIONS_API__: JSON.stringify(true),
```
Comment (MaxLeiter) [inline webpack.config-test.js:40]: I think I'd prefer to just hard code "true" or "false" here and avoid the JSON call

File: webpack.config.js [hunk 2]
```diff
@@ -88,6 +88,10 @@ const config = {
 	},
 	plugins: [
 		new VueLoaderPlugin(),
+		new webpack.DefinePlugin({
```
Comment (MaxLeiter) [inline webpack.config.js:91]: Same as last comment re JSON.stringify

**Review by MaxLeiter — CHANGES_REQUESTED** [review]

**Review by sha1sum — COMMENTED** [review]
The back button seems to not be functioning properly. I will look into this.

File: client/components/App.vue [hunk 3]
```diff
@@ -39,13 +41,29 @@ export default {
 		ConfirmDialog,
 		Mentions,
 	},
+	setup() {
```
Comment (MaxLeiter) [inline client/components/App.vue:44]: Can we and do we want to use https://vuejs.org/api/sfc-script-setup.html#script-setup?

File: client/components/InlineChannel.vue [hunk 4]
```diff
@@ -26,5 +26,14 @@ export default {
 			});
 		},
 	},
+	template: `
```
Comment (xPaw) [inline client/components/InlineChannel.vue:29]: Why the duplicated templated?

File: client/components/Username.vue [hunk 5]
```diff
@@ -49,5 +49,13 @@ export default {
 			});
 		},
 	},
+	template: `<span
```
Comment (xPaw) [inline client/components/Username.vue:52]: Duplicate template here too.

File: client/js/helpers/parse.js [hunk 6]
```diff
@@ -79,23 +80,20 @@ function parse(createElement, text, message = undefined, network = undefined) {
 	// arrays of objects containing start and end markers, as well as metadata
 	// depending on what was found (channel or link).
 	const channelPrefixes = network ? network.serverOptions.CHANTYPES : ["#", "&"];
-	const userModes = network?.serverOptions?.PREFIX.symbols || ["!", "@", "%", "+"];
```
Comment (xPaw) [inline client/js/helpers/parse.js:82]: Is this change necessary? `network?.` should be same as doing `network ? network...`

File: client/js/router.js [hunk 7]
```diff
@@ -107,12 +105,12 @@ router.beforeEach((to, from, next) => {
 	}
 
 	// Handle closing image viewer with the browser back button
-	if (!router.app.$refs.app) {
+	if (!router.app.$refs) {
```
Comment (xPaw) [inline client/js/router.js:108]: Looks suspicious

File: client/components/MessageTypes/whois.vue [hunk 8]
```diff
@@ -55,10 +55,10 @@
 			</template>
 
 			<template v-if="message.whois.special">
-				<template v-for="special in message.whois.special">
-					<dt :key="special">Special:</dt>
-					<dd :key="special">{{ special }}</dd>
-				</template>
+				<div v-for="special in message.whois.special" :key="special">
```
Comment (xPaw) [inline client/components/MessageTypes/whois.vue:58]: `div` is not necessary here, is it?

**Comment by xPaw** [comment]
In my branch I had to change v-slot for RevealPassword, is that no longer necessary? https://github.com/thelounge/thelounge/commit/2f8f7e76336c4f37b336e2d892c80ec9e23417c0

**Comment by MaxLeiter** [comment]
hey @sha1sum, are you still interested in taking this on? Now that 4.3.1 is out we can review + merge this if you can address our comments

**Comment by MaxLeiter** [comment]
Thanks for your efforts here @sha1sum! I’ve included some of your changes in #4559 and will be sure to credit you in the changelog / release notes 🙏

**Comment by sha1sum** [comment]
@MaxLeiter I really appreciate that. I somehow completely missed your question a couple of months ago (GitHub notifications can be noisy). Sorry about that. I hope that some of this helped!

**Pull request closed without merging**